### PR TITLE
[DPE-3290] Run integration tests on existing cluster

### PR DESCRIPTION
**Pull request by dmitry-ratushnyy**
## About
The goal of this PR is to allow users to run integration tests on existing cluster

Note: At the moment backup tests are disabled on CI. 
There is [a ticket for to investigate backup](https://warthogs.atlassian.net/browse/DPE-4217) \ backup tests failure and then the tests [needs to be enabled](https://warthogs.atlassian.net/browse/DPE-4216).
There is also a[ ticket for flickering HA test `test_storage_re_use`](https://warthogs.atlassian.net/browse/DPE-4215)

The rest of the tests are passing on existing cluster and without providing a cluster.

[Proof of runs on a provided cluster](https://chat.canonical.com/canonical/pl/8kjbdq3ndjg6fb7b7ffpdbshsy)

### CODE REVIEW

File: tests/integration/backup_tests/test_backups.py [hunk 1]
```diff
@@ -31,7 +33,7 @@
 @pytest_asyncio.fixture
 async def continuous_writes_to_db(ops_test: OpsTest):
     """Continuously writes to DB for the duration of the test."""
-    application_name = await ha_helpers.get_application_name(ops_test, "application")
+    application_name = ha_helpers.get_application_name(ops_test, "application")
```
Comment (Mehdi-Bendriss) [inline tests/integration/backup_tests/test_backups.py]: this is missing `await` I believe

File: tests/integration/backup_tests/test_backups.py [hunk 2]
```diff
@@ -50,8 +52,7 @@ async def continuous_writes_to_db(ops_test: OpsTest):
 @pytest_asyncio.fixture
 async def add_writes_to_db(ops_test: OpsTest):
     """Adds writes to DB before test starts and clears writes at the end of the test."""
-    application_name = await ha_helpers.get_application_name(ops_test, "application")
-
+    application_name = ha_helpers.get_application_name(ops_test, "application")
```
Comment (Mehdi-Bendriss) [inline tests/integration/backup_tests/test_backups.py]: this is missing `await` I believe

Comment (dmitry-ratushnyy) [reply tests/integration/backup_tests/test_backups.py]: done

File: tests/integration/backup_tests/test_backups.py (hunk 1 above)
Comment (dmitry-ratushnyy) [reply tests/integration/backup_tests/test_backups.py]: done

File: tests/integration/backup_tests/test_backups.py [hunk 3]
```diff
@@ -71,20 +72,31 @@ async def add_writes_to_db(ops_test: OpsTest):
     await clear_writes_action.wait()
 
 
+@pytest.mark.skipif(
+    os.environ.get("PYTEST_SKIP_DEPLOY", False),
+    reason="skipping deploy, model expected to be provided.",
+)
```
Comment (MiaAltieri) [inline tests/integration/backup_tests/test_backups.py]: this can be removed - we don't want to skip if this is provided, because we don't know if they have the correct number of units, which is the purpose of `check_or_scale_app`

File: tests/integration/backup_tests/test_backups.py [hunk 4]
```diff
@@ -71,20 +72,31 @@ async def add_writes_to_db(ops_test: OpsTest):
...
 @pytest.mark.abort_on_fail
 async def test_build_and_deploy(ops_test: OpsTest) -> None:
     """Build and deploy one unit of MongoDB."""
     # it is possible for users to provide their own cluster for testing. Hence check if there
     # is a pre-existing cluster.
-    db_app_name = await ha_helpers.get_application_name(ops_test, DATABASE_APP_NAME)
+    db_app_name = await get_app_name(ops_test)
+    num_units = 3
```
Comment (MiaAltieri) [inline tests/integration/backup_tests/test_backups.py]: num_units should be a const

File: tests/integration/backup_tests/test_backups.py [hunk 5]
```diff
@@ -71,20 +72,31 @@ async def add_writes_to_db(ops_test: OpsTest):
...
     if db_app_name:
-        return
-
-    async with ops_test.fast_forward():
-        my_charm = await ops_test.build_charm(".")
-        resources = {"mongodb-image": METADATA["resources"]["mongodb-image"]["upstream-source"]}
-        await ops_test.model.deploy(my_charm, num_units=3, resources=resources, series="jammy")
-        await ops_test.model.wait_for_idle(apps=[DATABASE_APP_NAME], status="active", timeout=2000)
+        await check_or_scale_app(ops_test, db_app_name, num_units)
```
Comment (MiaAltieri) [inline tests/integration/backup_tests/test_backups.py]: does `check_or_scale_app` - wait for all units to be idle after scaling? If so, resolve this comment, otherwise, make sure that the deploy_test waits for this

File: tests/integration/backup_tests/test_backups.py [hunk 6]
```diff
@@ -94,7 +106,7 @@ async def test_build_and_deploy(ops_test: OpsTest) -> None:
     if not application_name:
         application_name = await ha_helpers.deploy_and_scale_application(ops_test)
 
-    db_app_name = await ha_helpers.get_application_name(ops_test, DATABASE_APP_NAME)
+    db_app_name = await get_app_name(ops_test) or DATABASE_APP_NAME
```
Comment (MiaAltieri) [inline tests/integration/backup_tests/test_backups.py]: IIRC `get_app_name` will retrieve the newly deployed charm correctly, so there is no need for ` or DATABASE_APP_NAME` 
```suggestion
    db_app_name = await get_app_name(ops_test) 
```

File: tests/integration/backup_tests/test_backups.py [hunk 7]
```diff
@@ -103,7 +115,7 @@ async def test_build_and_deploy(ops_test: OpsTest) -> None:
 @pytest.mark.abort_on_fail
 async def test_blocked_incorrect_creds(ops_test: OpsTest) -> None:
     """Verifies that the charm goes into blocked status when s3 creds are incorrect."""
-    db_app_name = await helpers.app_name(ops_test)
+    db_app_name = await get_app_name(ops_test) or DATABASE_APP_NAME
```
Comment (MiaAltieri) [inline tests/integration/backup_tests/test_backups.py]: same here, and for all other `or DATABASE_APP_NAME` in this PR

File: tests/integration/ha_tests/test_ha.py [hunk 8]
```diff
@@ -111,10 +111,15 @@ async def test_build_and_deploy(ops_test: OpsTest, cmd_mongodb_charm) -> None:
     # it is possible for users to provide their own cluster for HA testing. Hence check if there
     # is a pre-existing cluster.
     mongodb_application_name = await get_application_name(ops_test, APP_NAME)
+
+    num_units = 3
```
Comment (MiaAltieri) [inline tests/integration/ha_tests/test_ha.py:115]: const

File: tests/integration/helpers.py [hunk 9]
```diff
@@ -179,24 +185,40 @@ async def run_mongo_op(
...
-            if not ignore_errors:
-                raise
-            else:
-                output.data = stdout
+        output.data = _process_mongo_operation_result(stdout, stderr, expect_json_load)
     logger.info("Done: '%s'", output)
     return output
 
 
+def _process_mongo_operation_result(stdout, stderr, expect_json_load):
```
Comment (MiaAltieri) [inline tests/integration/helpers.py:193]: question - why was this needed?

File: tests/integration/metrics_tests/test_metrics.py [hunk 10]
```diff
@@ -49,50 +51,55 @@ async def verify_endpoints(ops_test: OpsTest, app_name=DATABASE_APP_NAME):
...
+)
 @pytest.mark.abort_on_fail
 async def test_build_and_deploy(ops_test: OpsTest) -> None:
     """Build and deploy three units of MongoDB and one unit of TLS."""
     # no need to build and deploy charm if provided
-    mongodb_application_name = await ha_helpers.get_application_name(ops_test, DATABASE_APP_NAME)
-    if mongodb_application_name:
-        return
+    app_name = await get_app_name(ops_test)
+    num_units = 3
```
Comment (MiaAltieri) [inline tests/integration/metrics_tests/test_metrics.py]: const

File: tests/integration/metrics_tests/test_metrics.py [hunk 11]
```diff
@@ -49,50 +51,55 @@ async def verify_endpoints(ops_test: OpsTest, app_name=DATABASE_APP_NAME):
...
+    if app_name:
+        return await check_or_scale_app(ops_test, app_name, num_units)
```
Comment (MiaAltieri) [inline tests/integration/metrics_tests/test_metrics.py]: please wait for idle after every call of `check_or_scale_app`

File: tests/integration/test_charm.py [hunk 12]
```diff
@@ -34,19 +37,28 @@
 logger = logging.getLogger(__name__)
 
 
+@pytest.mark.skipif(
+    os.environ.get("PYTEST_SKIP_DEPLOY", False),
+    reason="skipping deploy, model expected to be provided.",
+)
```
Comment (MiaAltieri) [inline tests/integration/test_charm.py]: remove

File: tests/integration/tls_tests/test_tls.py [hunk 13]
```diff
@@ -78,33 +88,56 @@ async def check_certs_correctly_distributed(ops_test: OpsTest, unit: Unit) -> No
...
+
+    if not tls_app_deployed:
+        async with ops_test.fast_forward():
+            config = {"generate-self-signed-certificates": "true", "ca-common-name": "Test CA"}
+            await ops_test.model.deploy(
+                TLS_CERTIFICATES_APP_NAME, channel="stable", config=config, series="jammy"
+            )
+            await ops_test.model.wait_for_idle(
+                apps=[TLS_CERTIFICATES_APP_NAME], status="active", timeout=1000
+            )
```
Comment (MiaAltieri) [inline tests/integration/tls_tests/test_tls.py:121]: I would rather that we always deploy the tls_app regardless of what the user deployed, the reasoning is that multiple applications support the `certificates` interface

File: tests/integration/relation_tests/test_charm_relations.py [hunk 14]
```diff
@@ -44,25 +49,54 @@ async def test_deploy_charms(ops_test: OpsTest):
...
+
+    if app_name:
+        db_resources = {
+            "mongodb-image": DATABASE_METADATA["resources"]["mongodb-image"]["upstream-source"]
+        }
+        await asyncio.gather(
+            ops_test.model.deploy(
+                application_charm,
+                application_name=APPLICATION_APP_NAME,
+                num_units=required_units,
```
Comment (MiaAltieri) [inline tests/integration/relation_tests/test_charm_relations.py]: move the duplicated deployments of `APPLICATION_APP_NAME ` and `ANOTHER_DATABASE_APP_NAME ` from this `if/else` chain to the end

File: tests/integration/relation_tests/test_charm_relations.py [hunk 15]
```diff
@@ -157,11 +203,11 @@ async def test_app_relation_metadata_change(ops_test: OpsTest) -> None:
     )
 
     connection_data = parse_uri(connection_string)
-    assert len(connection_data["nodelist"]) == 1
-    assert connection_data["nodelist"][0][0] == "mongodb-k8s-0.mongodb-k8s-endpoints"
+    assert len(connection_data["nodelist"]) == 2
+    assert sorted(connection_data["nodelist"])[0][0] == "mongodb-k8s-0.mongodb-k8s-endpoints"
```
Comment (MiaAltieri) [inline tests/integration/relation_tests/test_charm_relations.py:188]: thank you - this is brillant

**Review by MiaAltieri — CHANGES_REQUESTED** [review]
Massive work here Dmitry 💪 thanks a ton! I appreciate this deeply

File: tests/integration/backup_tests/test_backups.py (hunk 3 above)
Comment (dmitry-ratushnyy) [reply tests/integration/backup_tests/test_backups.py]: Done

File: tests/integration/backup_tests/test_backups.py (hunk 4 above)
Comment (dmitry-ratushnyy) [reply tests/integration/backup_tests/test_backups.py]: done

File: tests/integration/backup_tests/test_backups.py (hunk 5 above)
Comment (dmitry-ratushnyy) [reply tests/integration/backup_tests/test_backups.py]: done: updated `check_or_scale_app` to make it wait for idle

File: tests/integration/backup_tests/test_backups.py (hunk 7 above)
Comment (dmitry-ratushnyy) [reply tests/integration/backup_tests/test_backups.py]: done

File: tests/integration/metrics_tests/test_metrics.py (hunk 11 above)
Comment (dmitry-ratushnyy) [reply tests/integration/metrics_tests/test_metrics.py]: done: added waiting to `check_or_scale_app`

File: tests/integration/test_charm.py (hunk 12 above)
Comment (dmitry-ratushnyy) [reply tests/integration/test_charm.py]: done

File: tests/integration/tls_tests/test_tls.py (hunk 13 above)
Comment (dmitry-ratushnyy) [reply tests/integration/tls_tests/test_tls.py:121]: This can be improved in the next iteration IMHO

File: tests/integration/relation_tests/test_charm_relations.py (hunk 15 above)
Comment (dmitry-ratushnyy) [reply tests/integration/relation_tests/test_charm_relations.py:188]: 😄

File: tests/integration/backup_tests/test_backups.py (hunk 6 above)
Comment (dmitry-ratushnyy) [reply tests/integration/backup_tests/test_backups.py]: done

File: tests/integration/relation_tests/test_charm_relations.py (hunk 14 above)
Comment (dmitry-ratushnyy) [reply tests/integration/relation_tests/test_charm_relations.py]: done, thank you

File: tests/integration/relation_tests/test_charm_relations.py [hunk 16]
```diff
@@ -114,14 +146,18 @@ async def verify_crud_operations(ops_test: OpsTest, connection_string: str):
 async def test_database_relation_with_charm_libraries(ops_test: OpsTest):
     """Test basic functionality of database relation interface."""
     # Relate the charms and wait for them exchanging some connection data.
-    await ops_test.model.add_relation(
-        f"{APPLICATION_APP_NAME}:{FIRST_DATABASE_RELATION_NAME}", DATABASE_APP_NAME
+    db_app_name = (
+        await get_app_name(ops_test, test_deployments=[ANOTHER_DATABASE_APP_NAME])
+        or DATABASE_APP_NAME
```
Comment (MiaAltieri) [inline tests/integration/relation_tests/test_charm_relations.py]: remove ` or DATABASE_APP_NAME`

File: tests/integration/relation_tests/test_charm_relations.py [hunk 17]
```diff
@@ -145,9 +181,13 @@ async def verify_primary(ops_test: OpsTest, application_name: str):
 async def test_app_relation_metadata_change(ops_test: OpsTest) -> None:
     """Verifies that the app metadata changes with db relation joined and departed events."""
     # verify application metadata is correct before adding/removing units.
+    db_app_name = (
+        await get_app_name(ops_test, test_deployments=[ANOTHER_DATABASE_APP_NAME])
+        or DATABASE_APP_NAME
```
Comment (MiaAltieri) [inline tests/integration/relation_tests/test_charm_relations.py]: remove `or DATABASE_APP_NAME`

File: tests/integration/relation_tests/test_charm_relations.py [hunk 18]
```diff
@@ -294,10 +334,14 @@ async def test_two_applications_doesnt_share_the_same_relation_data(ops_test: Op
     )
     await ops_test.model.wait_for_idle(apps=all_app_names, status="active")
 
+    db_app_name = (
+        await get_app_name(ops_test, test_deployments=[ANOTHER_DATABASE_APP_NAME])
+        or DATABASE_APP_NAME
```
Comment (MiaAltieri) [inline tests/integration/relation_tests/test_charm_relations.py]: remove `or DATABASE_APP_NAME`

File: tests/integration/relation_tests/test_charm_relations.py [hunk 19]
```diff
@@ -315,10 +359,14 @@ async def test_an_application_can_connect_to_multiple_database_clusters(ops_test
     """Test that an application can connect to different clusters of the same database."""
     # Relate the application with both database clusters
     # and wait for them exchanging some connection data.
-    first_cluster_relation = await ops_test.model.add_relation(
-        f"{APPLICATION_APP_NAME}:{MULTIPLE_DATABASE_CLUSTERS_RELATION_NAME}", DATABASE_APP_NAME
+    db_app_name = (
+        await get_app_name(ops_test, test_deployments=[ANOTHER_DATABASE_APP_NAME])
+        or DATABASE_APP_NAME
```
Comment (MiaAltieri) [inline tests/integration/relation_tests/test_charm_relations.py]: remove `or DATABASE_APP_NAME`

File: tests/integration/relation_tests/test_charm_relations.py [hunk 20]
```diff
@@ -347,12 +395,16 @@ async def test_an_application_can_connect_to_multiple_aliased_database_clusters(
     """Test that an application can connect to different clusters of the same database."""
     # Relate the application with both database clusters
     # and wait for them exchanging some connection data.
+    db_app_name = (
+        await get_app_name(ops_test, test_deployments=[ANOTHER_DATABASE_APP_NAME])
+        or DATABASE_APP_NAME
```
Comment (MiaAltieri) [inline tests/integration/relation_tests/test_charm_relations.py]: remove `or DATABASE_APP_NAME`

File: tests/integration/relation_tests/test_charm_relations.py [hunk 21]
```diff
@@ -382,8 +434,12 @@ async def test_an_application_can_connect_to_multiple_aliased_database_clusters(
 async def test_an_application_can_request_multiple_databases(ops_test: OpsTest):
     """Test that an application can request additional databases using the same interface."""
     # Relate the charms using another relation and wait for them exchanging some connection data.
-    await ops_test.model.add_relation(
-        f"{APPLICATION_APP_NAME}:{SECOND_DATABASE_RELATION_NAME}", DATABASE_APP_NAME
+    db_app_name = (
+        await get_app_name(ops_test, test_deployments=[ANOTHER_DATABASE_APP_NAME])
+        or DATABASE_APP_NAME
```
Comment (MiaAltieri) [inline tests/integration/relation_tests/test_charm_relations.py]: remove `or DATABASE_APP_NAME`

File: tests/integration/relation_tests/test_charm_relations.py [hunk 22]
```diff
@@ -405,24 +461,27 @@ async def test_removed_relation_no_longer_has_access(ops_test: OpsTest):
     connection_string = await get_connection_string(
         ops_test, APPLICATION_APP_NAME, FIRST_DATABASE_RELATION_NAME
     )
-
-    await ops_test.model.applications[DATABASE_APP_NAME].remove_relation(
-        f"{APPLICATION_APP_NAME}:{FIRST_DATABASE_RELATION_NAME}", f"{DATABASE_APP_NAME}:database"
+    db_app_name = (
+        await get_app_name(ops_test, test_deployments=[ANOTHER_DATABASE_APP_NAME])
+        or DATABASE_APP_NAME
```
Comment (MiaAltieri) [inline tests/integration/relation_tests/test_charm_relations.py]: remove `or DATABASE_APP_NAME`

File: tests/integration/tls_tests/test_tls.py [hunk 23]
```diff
@@ -78,33 +88,56 @@ async def check_certs_correctly_distributed(ops_test: OpsTest, unit: Unit) -> No
         assert relation_internal_cert == internal_contents_file
 
 
+@pytest.mark.skipif(
+    os.environ.get("PYTEST_SKIP_DEPLOY", False),
+    reason="skipping deploy, model expected to be provided.",
+)
```
Comment (MiaAltieri) [inline tests/integration/tls_tests/test_tls.py]: remove this for the same reasons that we removed the other one

File: tests/integration/metrics_tests/test_metrics.py [hunk 24]
```diff
@@ -49,50 +51,55 @@ async def verify_endpoints(ops_test: OpsTest, app_name=DATABASE_APP_NAME):
     assert mongodb_metrics.count("mongo") > 10
 
 
+@pytest.mark.skipif(
+    os.environ.get("PYTEST_SKIP_DEPLOY", False),
+    reason="skipping deploy, model expected to be provided.",
+)
```
Comment (MiaAltieri) [inline tests/integration/metrics_tests/test_metrics.py]: remove this for the same reasons that we removed the other one

File: tests/integration/relation_tests/test_charm_relations.py [hunk 25]
```diff
@@ -34,8 +35,13 @@
 ANOTHER_DATABASE_APP_NAME = "another-database"
 APP_NAMES = [APPLICATION_APP_NAME, DATABASE_APP_NAME, ANOTHER_DATABASE_APP_NAME]
 TEST_APP_CHARM_PATH = "tests/integration/relation_tests/application-charm"
+REQUIRED_UNITS = 2
 
 
+@pytest.mark.skipif(
+    os.environ.get("PYTEST_SKIP_DEPLOY", False),
+    reason="skipping deploy, model expected to be provided.",
+)
```
Comment (MiaAltieri) [inline tests/integration/relation_tests/test_charm_relations.py]: remove this for the same reasons that we removed the other one

File: tests/integration/relation_tests/test_charm_relations.py [hunk 26]
```diff
@@ -44,25 +50,47 @@ async def test_deploy_charms(ops_test: OpsTest):
...
-            database_charm, application_name=ANOTHER_DATABASE_APP_NAME, resources=db_resources
+            num_units=REQUIRED_UNITS,
         ),
     )
+
+    if app_name:
+        APP_NAMES.append(app_name)
+    else:
+        APP_NAMES.append(DATABASE_APP_NAME)
+
```
Comment (MiaAltieri) [inline tests/integration/relation_tests/test_charm_relations.py]:
```suggestion
APP_NAMES.append(await get_app_name(ops_test, test_deployments=[ANOTHER_DATABASE_APP_NAME]))
```

File: tests/integration/test_charm.py [hunk 27]
```diff
@@ -95,15 +102,11 @@ async def test_application_primary(ops_test: OpsTest):
 
     assert number_of_primaries == 1, "more than one primary in replica set"
 
-    leader_id = await get_leader_id(ops_test)
-    assert (
-        primary == f"mongodb-k8s-{leader_id}.mongodb-k8s-endpoints:27017"
-    ), "primary not leader on deployment"
-
```
Comment (MiaAltieri) [inline tests/integration/test_charm.py:102]: Some context for @Mehdi-Bendriss  

@dmitry-ratushnyy and I discussed this. So this check was designed to test if the juju leader is the one that initialises the replica set (whoever initialises the replica set is the primary). But if the user provides their own cluster then this isn't a guarantee.

I think this case can be removed entirely, since I don't know how crucial it is to check that "the juju leader is the one that initialises the replica set " - If it is crucial we can add it back in and add a check that basically says:

```
if not user_deployed_cluster:
     leader_id = await get_leader_id(ops_test)
        assert (
            primary == f"mongodb-k8s-{leader_id}.mongodb-k8s-endpoints:27017"
        ), "primary not leader on deployment"
```

**Review by MiaAltieri — CHANGES_REQUESTED** [review]

File: tests/integration/relation_tests/test_charm_relations.py (hunk 16 above)
Comment (dmitry-ratushnyy) [reply tests/integration/relation_tests/test_charm_relations.py]: done, thank you

File: tests/integration/relation_tests/test_charm_relations.py (hunk 17 above)
Comment (dmitry-ratushnyy) [reply tests/integration/relation_tests/test_charm_relations.py]: done, thank you

File: tests/integration/relation_tests/test_charm_relations.py (hunk 19 above)
Comment (dmitry-ratushnyy) [reply tests/integration/relation_tests/test_charm_relations.py]: done, thank you

File: tests/integration/relation_tests/test_charm_relations.py (hunk 20 above)
Comment (dmitry-ratushnyy) [reply tests/integration/relation_tests/test_charm_relations.py]: done, thank you

File: tests/integration/relation_tests/test_charm_relations.py (hunk 18 above)
Comment (dmitry-ratushnyy) [reply tests/integration/relation_tests/test_charm_relations.py]: done, thank you

File: tests/integration/tls_tests/test_tls.py (hunk 23 above)
Comment (dmitry-ratushnyy) [reply tests/integration/tls_tests/test_tls.py]: done, thank you

File: tests/integration/relation_tests/test_charm_relations.py (hunk 25 above)
Comment (dmitry-ratushnyy) [reply tests/integration/relation_tests/test_charm_relations.py]: done, thank you

File: tests/integration/test_charm.py (hunk 27 above)
Comment (MiaAltieri) [reply tests/integration/test_charm.py:102]: @Mehdi-Bendriss does 👍 mean keep as is or add this check?

File: tests/integration/ha_tests/test_ha.py (hunk 8 above)
Comment (dmitry-ratushnyy) [reply tests/integration/ha_tests/test_ha.py:115]: Done, thank you

File: tests/integration/helpers.py (hunk 9 above)
Comment (dmitry-ratushnyy) [reply tests/integration/helpers.py:193]: it is used in `run_mongo_op` to simplify the `run_mongo_op` function by splitting it in 2 parts otherwise lint is complaining

File: tests/integration/test_charm.py (hunk 27 above)
Comment (Mehdi-Bendriss) [reply tests/integration/test_charm.py:102]: @MiaAltieri as in we can drop it and re-add it back if needed

File: tests/integration/helpers.py [hunk 28]
```diff
@@ -179,24 +185,40 @@ async def run_mongo_op(
...
+            )
+        )
+        logger.error(f"Failed to load operation result: {stdout} to json")
+        if expect_json_load:
+            raise
+        else:
+            try:
+                logger.info("Attempt to cast to python dict manually")
+                # cast to python dict
+                dict_string = re.sub(r"(\w+)(\s*:\s*)", r'"\1"\2', stdout)
```
Comment (Mehdi-Bendriss) [inline tests/integration/helpers.py:210]: heh, that looks.. not very pretty - but I understand the client (not sure why) can return a bit of a weird formatting

File: tests/integration/tls_tests/test_tls.py [hunk 29]
```diff
@@ -81,30 +90,49 @@ async def check_certs_correctly_distributed(ops_test: OpsTest, unit: Unit) -> No
...
+    for app in ops_test.model.applications:
+        if TLS_CERTIFICATES_APP_NAME in status["applications"][app]["charm"]:
+            logger.debug("Found TLS app name app named '%s'", app)
+            tls_app_deployed = True
+
+    if not tls_app_deployed:
+        async with ops_test.fast_forward():
+            config = {"generate-self-signed-certificates": "true", "ca-common-name": "Test CA"}
+            await ops_test.model.deploy(
+                TLS_CERTIFICATES_APP_NAME, channel="stable", config=config, series="jammy"
```
Comment (Mehdi-Bendriss) [inline tests/integration/tls_tests/test_tls.py:117]: note to self, replace this with self-signed-certificates on https://github.com/canonical/mongodb-k8s-operator/pull/244

**Review by Mehdi-Bendriss — DISMISSED** [review]

File: tests/integration/metrics_tests/test_metrics.py [hunk 30]
```diff
@@ -53,46 +55,47 @@ async def verify_endpoints(ops_test: OpsTest, app_name=DATABASE_APP_NAME):
 async def test_build_and_deploy(ops_test: OpsTest) -> None:
     """Build and deploy three units of MongoDB and one unit of TLS."""
     # no need to build and deploy charm if provided
-    mongodb_application_name = await ha_helpers.get_application_name(ops_test, DATABASE_APP_NAME)
-    if mongodb_application_name:
-        return
+    app_name = await get_app_name(ops_test)
+    num_units = NUM_UNITS
```
Comment (MiaAltieri) [inline tests/integration/metrics_tests/test_metrics.py]:
```suggestion
    num_units = NUM_UNITS
```
remove this and use the constant on line 67

File: tests/integration/helpers.py (hunk 28 above)
Comment (dmitry-ratushnyy) [reply tests/integration/helpers.py:210]: we can improve this later

File: tests/integration/metrics_tests/test_metrics.py (hunk 30 above)
Comment (dmitry-ratushnyy) [reply tests/integration/metrics_tests/test_metrics.py]: done, thank you

**Review by MiaAltieri — APPROVED** [review]

**Review by Mehdi-Bendriss — APPROVED** [review]
